### PR TITLE
API for runtime warnings

### PR DESCRIPTION
**Pull request by exyi**
RuntimeWarningCollector is a service, where a new warning will can be
added using the `Warn` method. Warning is an instance of
DotvvmRuntimeWarning, contains a message and may contain a reference to
an exception and a bindable object.

By default, warnings are collected in Debug mode. They are displayed in
browser JS console and forwarded to Asp.Net Core logging.

IDotvvmWarningSink is extensibility point for people that want to
collect these warnings on their own.

As a demo, HtmlWriter reports warnings when elements that should be
always self-closing contain some content, as the browser will interpret
the content as sibling.

related to #745, but does not close it since this PR introduces only the infrastructure.

Also note that I had to move rendering of resources in postback after rendering of `PostBack.Update` controls. Since we support resource registration in the `Render` phase, this would need to be done anyway. I have altered `IHtmlWriter` and `IViewModelSerializer`, but these interfaces are so bound to the implementations in framework, that I don't believe somebody has implemented them...

The public API looks like this:

```C#
namespace DotVVM.Framework.Runtime
{
    public interface IDotvvmWarningSink
    {
        void RuntimeWarning(DotvvmRuntimeWarning warning);
    }
    public class DotvvmRuntimeWarning
    {
        public DotvvmRuntimeWarning(string message, Exception relatedException = null, DotvvmBindableObject relatedControl = null);

        public string Message { get; }
        public Exception RelatedException { get; }
        public DotvvmBindableObject RelatedControl { get; }

        public override string ToString();
    }

    public class RuntimeWarningCollector
    {
        public bool Enabled { get; }

        public RuntimeWarningCollector(IEnumerable<IDotvvmWarningSink> sinks, DotvvmConfiguration config);

        public void Warn(DotvvmRuntimeWarning warning);


        public List<DotvvmRuntimeWarning> GetWarnings() => warnings.ToList();
    }
}
```

### CODE REVIEW

File: src/DotVVM.Framework/Controls/HtmlWriter.cs [hunk 1]
```diff
@@ -191,6 +202,9 @@ public void RenderSelfClosingTag(string name)
         {
             RenderBeginTagCore(name);
             writer.Write("/>");
+
+            if (this.enableWarnings && !IsSelfClosing(name))
+                Warn($"Element {name} is not self closing but is rendered as so. It may interpreted as a start tag without the end tag by the browsers.");
```
Comment (ENgateman) [inline src/DotVVM.Framework/Controls/HtmlWriter.cs]: You're missing the "be" in "be interpreted".

If start tag has indefinite article, end tag should most likely have it as well: "an end tag".

Also, the tiniest of nitpicks: "self-closing".

File: src/DotVVM.Framework/Controls/HtmlWriter.cs [hunk 2]
```diff
@@ -352,6 +366,9 @@ public void RenderEndTag()
                 writer.Write("</");
                 writer.Write(tag);
                 writer.Write(">");
+
+                if (this.enableWarnings && IsSelfClosing(tag))
+                    Warn($"Element {tag} is self closing but contains content. The browser may interpret the start tag as self-closing and put the 'content' into it's parent.");
```
Comment (ENgateman) [inline src/DotVVM.Framework/Controls/HtmlWriter.cs]: Come on, you even have it here (still looking at "self-closing").

~"it's parent"~ -> "its parent".

Comment (exyi) [reply src/DotVVM.Framework/Controls/HtmlWriter.cs]: Thanks, fixed that.

File: src/DotVVM.Framework/Controls/HtmlWriter.cs (hunk 1 above)
Comment (exyi) [reply src/DotVVM.Framework/Controls/HtmlWriter.cs]: Thanks, fixed that.

**Pull request merged**